issue #1482 initial contribution of production in-memory backend

dk-github opened the pull request:
We would like to contribute an improved implementation of in-memory backend for JanusGraph.

Please see issue #1482 for rationale and some comparative analysis vs existing "test-only" in-memory backend.

**_NOTE_**: This PR currently suggests adding the new backend rather than modifying the existing one, as it seemed cleaner and easier to compare performance within one codebase branch.
However, having read recent discussions about adding more backends into main repository vs maintenance costs, I am starting to think that actually modifying the existing in-memory backend could be a better idea.
This is because:

- the modifications are fairly straightforward and do not change the general structure of existing backend
- don’t add too much extra code or any external dependencies
- don’t require any additional documentation or setup instructions, at least initially
- all existing (and a few extra) integrity tests pass and using it in place of current version is unlikely to create any issues for current (test-only) uses
- the same backend will simply be fit for production use, no new backends to maintain

Thoughts?

dk-github commented:
> Committer of one or more commits is not listed as a CLA signer, either individual or as a member of an organization.

In fact I should be listed as per this: https://github.com/JanusGraph/legal/pull/128 - so not sure what it complains about...

pluradj commented:
The problem is likely that the `Author` email on your commit does not match the `Signed-off-by` email. You can follow these [instructions](https://github.com/JanusGraph/janusgraph/blob/master/CONTRIBUTING.md#configure-your-repo-to-match-the-cla) to amend it.

```
Author: Dmitry Kovalev <Dmitry.Kovalev@gs.com>
Date:   Fri Mar 1 13:56:37 2019 +0000

    initial contribution of CompactInMemoryStoreManager - an improved in-memory backend for JanusGraph, which offers significantly reduced memory footprint compared to "classic" one, and is suitable for production use
    
    Signed-off-by: Dmitry Kovalev <dk.global@gmail.com>
```

mbrukman commented:
Yes, there's a mismatch between Author and `Signed-off-by`, but interestingly enough, the DCO check passed, but it is the CLA check from @janusgraph-bot that failed, and it currently does not look at DCO (there's an outstanding feature request to add it, but since it's not yet there, we're using an external service).

There's potentially an issue because the Author and Committer are different here (though that should not invalidate the CLA check); I'm looking into this right now.

mbrukman commented:
I think I see what the problem is; the CLA and the commit are not using the same capitalization for the email address. I anticipated this would happen when I saw your CLA submitted with upper-cased first-name/last-name in your emails, as I suspected some folks would actually lowercase their emails (and some folks on that thread already did).

In fact, I anticipated it so well that I already implemented case-insensitive email matching for CLA authors and committers (see https://github.com/google/code-review-bot/pull/16); the issue here was that I hadn't actually rebuilt the binary used by @janusgraph-bot to include this latest change.

I just pushed a new build, so if my analysis is correct, the CLA issue should be resolved shortly.

dk-github commented:
It worked indeed, thanks @mbrukman! There were internal reasons for capitals in the emails - apologies for causing trouble. But as per e-mail standard the addresses should be case-insensitive so IMO it was a generally useful fix anyway.

dk-github commented:
just a gentle weekly reminder - would be grateful if someone could review/approve

porunov commented on janusgraph-test/src/test/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPageTest.java:
```diff
@@ -0,0 +1,139 @@
+    {
+        Entry[] entries = new Entry[]
+                {
+                    StaticArrayEntry.of(StaticArrayBuffer.of(COL_START.getBytes()), StaticArrayBuffer.of("valStart".getBytes())),
```
You can simplify the code if you make a separate method like
```
private Entry makeEntry(String something1, String something2){
    return StaticArrayEntry.of(StaticArrayBuffer.of(something1.getBytes()), StaticArrayBuffer.of(something2.getBytes()))
}
```
Then you can use:
```
new Entry[]{
  makeEntry(COL_START, "valStart"),
  makeEntry("01col1", "val1"),
  ...
}
```

dk-github replied:
@porunov thanks for your comments. I went through them and they look to be mostly "cosmetic" so far. I am happy to discuss them further, and regardless of the results of the discussion or my personal opinion, ultimately I am happy to make any change of this type if that is going to be the condition for approval. 

However I suggest that we return to these when/if the actual logic and performance are in fact reviewed and accepted in general.

porunov replied:
@dk-github Sure. I agree here with you

amcp replied:
same opinion with @porunov , will make this class easier to read

dk-github replied:
done

mbrukman commented:
@twilmes — since you've recently written a new FoundationDB backend, can you please take a look at this new in-memory backend contribution? Thanks!

mbrukman commented:
@amcp – since you've written a storage backend for JanusGraph, would you mind reviewing this PR, please?

amcp commented:
First, I am reading the PR and I think that given that the data structure used in InMemoryKCVStore remained the same, this PR does well and does not create a new storage backend. I will dig into the details now.

dk-github commented:
@amcp please let me know if there are any questions or if there's anything I can do to help move it forward.

@FlorianHockmann you mentioned in a separate discussion that you might review this after you are finished with Tinkerpop upgrade - do you think you would be able to do it now?

FlorianHockmann commented:
@dk-github Thanks for the reminder. I want to review this PR, but I'm concentrating my efforts for JanusGraph currently on the 0.4.0 release as our last release from `master` (which means bigger features) was in October.
I think we can get this PR properly reviewed after that release and the same applies basically to the other big PRs we currently have open (#1437, #1591, and #1521).
We'll then have to decide whether this a breaking change or not. If not, we can just release it with 0.4.1 and if it's a breaking change, then we'll have to create a `0..4` branch so we can use `master` for development of 0.5.0, but that's also not a big problem.

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,6 +124,429 @@ public void close() throws BackendException {
         kcv.clear();
     }
 
+    public static class FragmentationReport {
```
Why is this class static?

dk-github replied:
not anymore

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPageUtils.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * This is a collection of static methods to help read a BufferPage from a stream it was previously dumped to
+ */
+public class BufferPageUtils {
```
This class should be final.

dk-github replied:
ok, done...

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPageUtils.java:
```diff
@@ -0,0 +1,78 @@
+    public static SharedEntryBuffer readFrom(DataInputStream in) throws IOException {
+        int numPages = in.readInt();
+
+        //System.out.println("number of pages in column store is " + numPages);
```
Could you remove this or use a logger with the level trace?

dk-github replied:
used logger

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,6 +124,429 @@ public void close() throws BackendException {
+    }
+
+    private void dumpChunk(Path filePath, List<Map.Entry<StaticBuffer, ColumnValueStore>> chunk) {
+        //System.out.println("number of column stores in chunk " + filePath + ": " + chunk.size() + " " + Thread.currentThread().getName());
```
Same here in all following places.

dk-github replied:
used logger

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/MultiPageEntryBuffer.java:
```diff
@@ -0,0 +1,399 @@
+    }
+
+    @Override
+    public FragmentationReport createFragmentationReport(int maxPageSize) {
```
To create FragmentationReport, it would be good to use a builder.

dk-github replied:
done, although it is not intended to ever be constructed by user code, and basically is likely to only be ever constructed in one place, so not so much scope for using this builder...

farodin91 commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/SharedEntryBuffer.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    class FragmentationReport {
```
You have already a class with the same name in the InMemoryKeyColumnValueStore. Would be possible to merge these class?

dk-github replied:
> Can you move it into a separate file?

done, although IMHO it was neater to have it as an inner class

dk-github replied:
> You have already a class with the same name in the InMemoryKeyColumnValueStore. Would be possible to merge these class?

they are substantially different - one is fragmentation report for a single page, another is for the whole store. both are now top-level classes with different names

dk-github commented:
> I added some comments. It would be good to split up janusgraph-core, so i would prefer to add a extra module and remove the complete inmemory backend from janusgraph-core.

Thank you for the comments @farodin91 , and apologies for the late reply.

As far as I can see, all of them are again of cosmetic/code style/refactoring suggestion type. As mentioned before, I am happy to address them all, but before spending time on them I would really like to be sure that maintainers of this project are actually happy with the design and the added value vs any concerns for this contribution, and would be in position to accept it if all the above "technical" comments are addressed.

Taking your review as example - can I assume that lack of any comments from you on the design/performance or any other concerns means that you have looked through these and are happy to approve it, provided that your comments above are addressed?.

Many thanks,
Dmitry

dk-github commented:
Hi All,

thank you for all the comments so far on the issue, PR and in the dev maillist. Unfortunately I was not able to obtain from these comments a clear understanding of how to proceed with this contribution, and there doesn't seem to be a lot of traffic on it in general. 

**So I have collated a few options based on the feedback so far, and I would like to ask everyone, especially the previous commenters and also preferably active developers/TSC members to look through them and cast their "vote" on the issue page:**

https://github.com/JanusGraph/janusgraph/issues/1482#issuecomment-536026375

Hopefully the fact that one just has to choose an option will increase participation :), but if you have more to say on the matter than just choose from the options, then by all means please do so.

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/ColumnValueStore.java:
```diff
@@ -33,26 +35,31 @@
 class ColumnValueStore {
 
-    private static final double SIZE_THRESHOLD = 0.66;
+    public static final int DEF_PAGE_SIZE = 500;
```
why public? if referred to in tests, make private or package private. private preferred so you can detect when the value is changed in source code here.

dk-github replied:
done

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -24,24 +25,43 @@
+import static org.janusgraph.diskstorage.StaticBuffer.ARRAY_FACTORY;
 
 /**
  * An in-memory implementation of {@link KeyColumnValueStore}.
```
i think you should update this comment. you changed the implementation.

dk-github replied:
done

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,6 +124,429 @@ public void close() throws BackendException {
+            Path filePath = Paths.get(storePath.toString(), getName() + "_" + i);
+
+            return parallelOperationsExecutor.submit(() -> dumpChunk(filePath, chunks.get(i)));
+        }).collect(Collectors.toList())
```
why is this collect operation necessary? you stream right after it,

dk-github replied:
it is necessary to ensure that the preceding map() is executed eagerly and all tasks are actually submitted to execute in parallel.
without the collect, it would execute lazily, submit tasks one by one and immediately block until it is done - so beating the purpose of executing them on the thread pool in the first place... 
will add a comment to this effect

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,6 +124,429 @@ public void close() throws BackendException {
+                sbKcv.dumpTo(out);
+            }
+
+            //System.out.println("finished writing chunk " + filePath + " " + Thread.currentThread().getName());
```
trace

dk-github replied:
done (debug)

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryStoreManager.java:
```diff
@@ -132,6 +139,59 @@ public String getName() {
+        if (rollbackIfFailed) {
+            //NOTE: if rollbackIfFailed is true, we clear current contents of stores only AFTER successful loading,
+            // thus having the option to get back to old data if load failed, but we then require 2x memory to hold both old and new data
+            //so the risk of OOM increases
```
minor add leading space

dk-github replied:
done

amcp commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/MultiPageEntryBuffer.java:
```diff
@@ -0,0 +1,399 @@
+
+        int numEntries = 0;
+
+        for (BufferPage p : pages) {
```
return pages.stream().map(BufferPage::numEntries).collectsum() or something similarshould be possible here.

dk-github replied:
numEntries() is on "critical path" in several places such as in getPageIndex(), and using stream for something as trivial as a sum is probably significantly more overhead than using good old cycle... 
I agree that in a case where we would do this sum once in a while, it would still be arguably worth doing stream just for readability. And I confess that I didn't actually profile to see if there will be any "noticeable(tm)" difference between using stream or not. But still, unless you insist, I would leave this as is. 
Please let me know if you still have concerns

amcp commented on janusgraph-test/src/test/java/org/janusgraph/diskstorage/inmemory/ClassicVsCompactComparisonTest.java:
```diff
@@ -0,0 +1,84 @@
+
+/**
+ * This is just a simple test program to illustrate comparative performance and memory footprint (as captured via JMC)
+ * between "classic" InMemory backend and the new improved one
```
put the test results here, before and after making the change. maybe this class should be a different commit and should precede this PR

dk-github replied:
Thinking about it, it probably doesn't belong in the actual JanusGraph code at all, as it is not a test we want to run as part of our build etc - it is just a program that illustrates the difference between "old" inmemory backend and "new" one. 
I guess I will add it as a code snippet or a mini-project in my private repo and link here?

amcp commented on janusgraph-test/src/test/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/ColumnValueStoreTest.java:
```diff
@@ -0,0 +1,529 @@
+ * This is a unit test for custom override of JanusGraph ColumnValueStore. Required to verify the added logic and
+ * the correctness of its integration with standard logic
+ */
+public class ColumnValueStoreTest
```
would rename this to InMemoryColumnValueStoreTest

dk-github replied:
done. together with the class it tests...

amcp commented:
I would consider some refactoring to address complexity issues reported by codacy. and definitely please do make the travis ci build pass.  @dk-github

dk-github commented:
> and definitely please do make the travis ci build pass

it was passing before, and the last build failed in some irrelevant CQL test:
[ERROR] Failures: 
[ERROR]   CQLGraphCacheTest>JanusGraphTest.testVertexTTLWithCompositeIndex:5362 expected: not <null>

but I will of course make sure that it is still green after addressing all the review comments (and rebasing it to the new location after#1864 is merged

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,292 @@
+        //NOTE: instead of constructing all Entries and then rebuilding offsetIndex and rawData,
+        // it could use arraycopy to copy parts of existing buffers into new ones
+        // it is possible to do in one pass, but rather complex as we'd need to keep track of current offset in new buffer,
+        // current page and offset within that page etc. In practice, this is required seldom enough to create any noticeable deficiencies.
```
for the purposes of this comment, I defined "noticeable" as "visible in the hot methods or anywhere in the captured stack traces during performance profiling of a real production usage pattern" :)

Basically, I it is probably possible to come up with a pattern of writes/deletes of sparsely distributed keys to the graph which will make the performance of this merge operation significant, but that must be a LOT of  merges. In reality, people mostly just add to the graph, and rarely delete from it, and the keys seem to be mostly continuous. 

So it just felt like a premature optimisation which is complex to implement but provides unclear advantages.

FlorianHockmann commented:
> it was passing before, and the last build failed in some irrelevant CQL test:
> [ERROR] Failures:
> [ERROR] CQLGraphCacheTest>JanusGraphTest.testVertexTTLWithCompositeIndex:5362 expected: not 

This is a known flaky test: #1464
For flaky tests, you can simply restart the Travis job and see if it passes in the second run. (That's obviously not a good solution, so we should tackle these flaky tests, but it at least doesn't stop PRs from being merged.)

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,292 @@
+            //NOTE: we used to do getColumn(mid).compareTo(column), where compareTo internally asserts that column is a StaticARRAYBuffer, so
+            // apparently we don't have to worry about column name being in anything but array-backed buffer
+            //So we can compare relevant array parts directly, avoiding the churn of new StaticArrayBuffers created by getColumn()
+            assert column instanceof StaticArrayBuffer;
```
done (using Preconditions.checkArgument which is what is used throughout the rest of the code)

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,295 @@
+        this(array, 0, size);
+    }
+
+    BufferPage(final Entry[] array, final int start, final int end) {
```
moved the code to a static build method as per the first suggestion.
it is on critical path and creating a full "builder" class instance every time would result in a lot of churn

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,295 @@
+// Copyright 2017 JanusGraph Authors
```
done

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,295 @@
+        if (offsetIndex.length > 0) {
+            //NOTE: looks like there is no other way to write out an int[] without copying it?.. no IntBuffer.asByteBuffer(endianness) or smth...
+            // will rely on buffering here
+            for (int i : offsetIndex)
```
done

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,295 @@
+
+            out.writeInt(rawData.length);
+            out.write(rawData);
+            //System.out.println("page data length is " + getRawData().length + ", byte size is " + dataBuffer.limit());
```
changed to use logger

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,6 +124,429 @@ public void close() throws BackendException {
         kcv.clear();
     }
 
+    public static class FragmentationReport {
```
done, although IMHO it was neater to have it as an inner class

dk-github commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/SharedEntryBuffer.java:
```diff
@@ -0,0 +1,120 @@
+
+    void dumpTo(DataOutputStream out) throws IOException;
+
+    class MemoryEntryList extends ArrayList<Entry> implements EntryList {
```
done, although if I remember correctly it was inner class in the original implementation, and the idea for the pre-existing classes was to only change what necessary and leave most code as it was

dk-github commented:
@amcp @farodin91 @porunov I have addressed what I believe to be all of your previous comments - please can you have another look and resolve if you agree

I have also addressed such of the codacy issues which made some sense and were trivial to fix. I was planning to see if I can address some of the rest of its complaints, but after my latest commit it miraculously says all is good?.. Has it changed its mind then? :)

dk-github commented:
Hi All, any chance previous reviewers can verify the requested changes please?

All comments addressed, all checks passed, build is green.

Some feedback on this before Christmas/NY would be a nice festive present, and would set us tentatively on course to merge the PR before the 1st anniversary of opening it :)

porunov commented:
@dk-github Hi, I would like to review this PR but I would like to make sure that all TinkerPop tests passes for this PR. Currently there is some problems with tinkerpop tests in the master branch (for some reason Cassandra docker container is not started during TinkerPop testing which causes a lot of test failures). I think we should fix it ASAP and then start to test this PR.
The PR to run tinkerpop tests can be found here: #1888
If you would like to start tinkerpop tests, just clone current master branch, merge that PR into the current master branch and execute `docker-build/docker-build.sh -p`. Notice, that the process is long and usually takes ~8-10 hours (on my machine) before the failure.

@farodin91 @FlorianHockmann do you know if there are other possibilities to test this tinkerpop tests for this PR?

farodin91 commented:
@porunov @dk-github the tinkerpop Tests are automatically executed with each Travis ci build for the immemory backend.

@dk-github i will try to review the pr next week.

dk-github commented:
@porunov @farodin91 thanks a lot

@porunov, for tinkerpop tests, have a look at this job from the last PR build:
https://travis-ci.org/JanusGraph/janusgraph/jobs/606886121?utm_medium=notification&utm_source=github_status

I am guessing ARGS='-Dtest.skip.tp=false' means they should have run, and you can search for familiar lines in the job log to spot-check your favourite TP tests in there

farodin91 commented:
> I am guessing ARGS='-Dtest.skip.tp=false' means they should have run, and you can search for familiar lines in the job log to spot-check your favourite TP tests in there

Correct.

farodin91 commented:
My PR #1902 reduces the errors in the inmemory tests.

porunov commented:
> the tinkerpop Tests are automatically executed with each Travis ci build for the immemory backend.

Yes, but sometimes we can break a thing which we thought to be unrelated. It is always good to execute all tests. Even so this PR is related only to in-memory backend, we can't be sure that the PR doesn't break something else.
I would like to execute all TP tests for big PRs to be sure it works correctly.
Currently I don't have a possibility to execute TP tests because of #1901 . 
Most likely this PR doesn't break any TP test (because in-memory TP tests passes), but again, I don't know how to verify other TP tests (which might seem to be unrelated) for this PR.

porunov commented:
As #1901 is fixed now, master branch passes all TP tests. I've started TP tests for master branch with this PR merged. Tests are started with `mvn clean install -Dtest.skip.tp=false -DskipTests=true`. I will provide a report lately, when tests are finished (ETA 10-15 hours).

porunov commented:
All TP tests passed for current master branch ([commit](https://github.com/JanusGraph/janusgraph/commit/05f4f9436cf80c1eb9f3bcbdc6bbe94160f11805)) with this PR merged into it.

porunov commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    public int byteSize() {
+        return rawData.length + offsetIndex.length * Integer.BYTES + 16;
```
(question) 
Do I understand it right that when we execute `byteSize` in `MultiPageEntryBuffer` it should return the sum of all buffers, where each bugger takes:
`offsetIndex.length * Integer.BYTES + 32 + rawData.length`? Is it a right formula?
If yes, then why do you add +16 here and +16 in `byteSize` method of `MultiPageEntryBuffer`? Why not to add +32 just here?

dk-github replied:
Usually most of the buffers in a typical Janusgraph store will be less than 500 entries (so one page), and thus using "multi-page" format for all of them would be a waste, since it requires more overhead bytes to store the ArrayList of pages, etc. So we start with SinglePageEntryBuffer, and only convert to MultiPage if we have to.

So for SinglePageEntryBuffer the total byte size will be exactly the same as for a single page, but for multi-page buffer the total byte size is (roughly) the sum of each page's size plus the size of the reference to the page from the list of pages (plus the overhead of the ArrayList internals).

porunov replied:
Thank you for the explanation @dk-github !

porunov commented:
@dk-github Thank you for this really great contribution! I've checked all the code as I could. I didn't go too deep into the logic, but the code looks good to me. I also checked my past comments and resolved them.
I have only a small one (possibly nitpick) question here https://github.com/JanusGraph/janusgraph/pull/1483#discussion_r360131160 
Also, could you please squash you commits into a single one?

Other then that, I don't have any more concerns.

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           

Complexity increasing per file
==============================
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java  22
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/SharedEntryBufferFragmentationReport.java  1
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBufferTest.java  1
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStore.java  9
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java  9
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MemoryEntryList.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManager.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java  5
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/TestPagedBufferColumnValueStore.java  1
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/SinglePageEntryBuffer.java  7
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/BufferPageTest.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStoreFragmentationReport.java  4
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java  14
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManagerTest.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStore.java  9
         

Clones added
============
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392&bid=11727944)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java:
```diff
@@ -0,0 +1,515 @@
+    public static final String VERY_END = "zz";
+
+    @Test
+    public void testRoundtrip()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBufferTest.java:
```diff
@@ -0,0 +1,76 @@
+import org.junit.jupiter.api.Test;
+
+import static org.janusgraph.diskstorage.inmemory.BufferPageTest.makeStaticBuffer;
+import static org.junit.jupiter.api.Assertions.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/BufferPageTest.java:
```diff
@@ -0,0 +1,148 @@
+import org.junit.jupiter.api.Test;
+
+import static org.janusgraph.diskstorage.inmemory.BufferPageUtils.buildFromEntryArray;
+import static org.junit.jupiter.api.Assertions.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java:
```diff
@@ -0,0 +1,406 @@
+    }
+
+    @Override
+    public void mutate(Entry[] add, Entry[] del, int maxPageSize) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'mutate(Entry, Entry, int)' has an NPath complexity of 4476](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'mutate(Entry, Entry, int)' has an NPath complexity of 4476](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManagerTest.java:
```diff
@@ -0,0 +1,142 @@
+
+import static org.janusgraph.diskstorage.inmemory.BufferPageTest.makeEntry;
+import static org.janusgraph.diskstorage.inmemory.BufferPageTest.makeStaticBuffer;
+import static org.junit.jupiter.api.Assertions.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java:
```diff
@@ -0,0 +1,515 @@
+
+import static org.janusgraph.diskstorage.inmemory.BufferPageTest.makeEntry;
+import static org.janusgraph.diskstorage.inmemory.BufferPageTest.makeStaticBuffer;
+import static org.junit.jupiter.api.Assertions.*;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/SharedEntryBufferFragmentationReport.java:
```diff
@@ -0,0 +1,111 @@
+        private int compressablePageCount;
+        private int achievablePageReduction;
+
+        public Builder() {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unnecessary constructors - the compiler will generate these for you](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManager.java:
```diff
@@ -132,6 +139,58 @@ public String getName() {
+            try {
+                newStores.put(storePath.getFileName().toString(), InMemoryKeyColumnValueStore.readFrom(storePath, storePath.getFileName().toString(), parallelOperationsExecutor));
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,17 +135,225 @@ public void close() throws BackendException {
+
+            return numKcvs;
+        } catch (Exception ex) {
+            throw new RuntimeException("Problem while reading chunk " + filePath + " of store " + store.getName(), ex);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,243 @@
+import org.janusgraph.diskstorage.util.StaticArrayBuffer;
+import org.janusgraph.diskstorage.util.StaticArrayEntry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.slf4j.LoggerFactory'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,243 @@
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.util.StaticArrayBuffer;
+import org.janusgraph.diskstorage.util.StaticArrayEntry;
+import org.slf4j.Logger;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.slf4j.Logger'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,243 @@
+        return newPages;
+    }
+
+    public List<BufferPage> merge(Entry[] add, int iaddp, int addLimit, Entry[] del, int idelp, int delLimit, int maxPageSize) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'merge(Entry, int, int, Entry, int, int, int)' has an NPath complexity of 728](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java:
```diff
@@ -0,0 +1,515 @@
+    }
+
+    @Test
+    public void testPagingAndFragmentation() throws TemporaryLockingException
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManagerTest.java:
```diff
@@ -0,0 +1,142 @@
+public class InMemoryStoreManagerTest
+{
+    @Test
+    public void testStoreCycle() throws Exception
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,17 +135,225 @@ public void close() throws BackendException {
+                log.debug("finished writing chunk " + filePath + " " + Thread.currentThread().getName());
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStore.java:
```diff
@@ -104,17 +135,225 @@ public void close() throws BackendException {
         kcv.clear();
     }
 
+    public InMemoryKeyColumnValueStoreFragmentationReport createFragmentationReport(StoreTransaction txh) throws BackendException {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'createFragmentationReport(StoreTransaction)' has an NPath complexity of 202](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'createFragmentationReport(StoreTransaction)' has an NPath complexity of 202](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java:
```diff
@@ -0,0 +1,406 @@
+    }
+
+    @Override
+    public EntryList getSlice(KeySliceQuery query) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'getSlice(KeySliceQuery)' has an NPath complexity of 2364](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'getSlice(KeySliceQuery)' has an NPath complexity of 2364](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

dk-github commented:
many thanks @porunov for taking time to revisit this PR

> I have only a small one (possibly nitpick) question here [#1483 (comment)](https://github.com/JanusGraph/janusgraph/pull/1483#discussion_r360131160)

replied to that comment, hopefully it answers the question

> Also, could you please squash you commits into a single one?

done. it has woken up Codacy again though. the same code which it deemed to be "ok" when it was in the form of 2 commits, it now again thinks is "not up to standard" :)

please let me know if you want me to address any of its comments specifically - preferably which ones exactly...
 
> Other then that, I don't have any more concerns.

great, does that mean that you are ready to approve?

porunov commented:
@amcp @farodin91 @FlorianHockmann will you be able to re-review this PR?

FlorianHockmann commented:
> @amcp @farodin91 @FlorianHockmann will you be able to re-review this PR?

I think @farodin91 already made a first pass through this and if he and you both accept this PR, then I'd say that we just merge it. I probably won't have the time to review this massive PR adequately. So, I don't want to block this.
Given that we have only recommended the in-memory backend for testing purposes until now, this can't break any production usages. Since all tests also still pass with this, the in-memory backend seems to still fulfil its original purpose. Therefore this PR isn't as critical as it may seem at first. (With critical I don't mean that it's not useful, only that it shouldn't break anything.)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/BufferPageTest.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManager.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java  14
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStore.java  9
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBufferTest.java  1
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/SinglePageEntryBuffer.java  7
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/SharedEntryBufferFragmentationReport.java  1
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java  9
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MemoryEntryList.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryKeyColumnValueStoreFragmentationReport.java  4
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStore.java  9
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java  5
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/TestPagedBufferColumnValueStore.java  1
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManagerTest.java  2
- janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java  22
         

Clones added
============
- janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392&bid=11727944)

FlorianHockmann commented on janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:
```diff
@@ -0,0 +1,241 @@
+        return newPages;
+    }
+
+    public List<BufferPage> merge(Entry[] add, int iaddp, int addLimit, Entry[] del, int idelp, int delLimit, int maxPageSize) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'merge(Entry, int, int, Entry, int, int, int)' has an NPath complexity of 728](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java:
```diff
@@ -0,0 +1,517 @@
+    public static final String VERY_END = "zz";
+
+    @Test
+    public void testRoundtrip()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryStoreManagerTest.java:
```diff
@@ -0,0 +1,142 @@
+public class InMemoryStoreManagerTest
+{
+    @Test
+    public void testStoreCycle() throws Exception
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

FlorianHockmann commented on janusgraph-inmemory/src/test/java/org/janusgraph/diskstorage/inmemory/InMemoryColumnValueStoreTest.java:
```diff
@@ -0,0 +1,517 @@
+    }
+
+    @Test
+    public void testPagingAndFragmentation() throws TemporaryLockingException
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)

dk-github commented:
Thank you @porunov @farodin91 . I have removed the unused imports/changes wildcards to specific imports - this seems to have pacified codacy a little bit.

Please let me know if I need to address anything else, or it is good to be merged.

porunov commented:
@dk-github I think you can merge this PR after Travis tests are finished

dk-github commented:
Ok, the build passed and I have merged it now. 

Many thanks to everyone who took time to review it, hope it will add some value to the project!

I will let it rest for a while, and then raise a small PR to update documentation about in-memory backend production usage and limitations. 

Also later this year may raise a number of small PRs with minor refactorings/improvements as/when we get to work on upgrading our system to janusgraph 0.5.0 and switching from (slightly more feature-rich) custom in-house backend to the contributed version.
